features/verifysession/impl/src/main/kotlin/io/element/android/features/verifysession/impl/VerifySelfSessionPresenter.kt

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2023 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.features.verifysession.impl
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.collectAsState
+import androidx.compose.runtime.derivedStateOf
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import io.element.android.libraries.architecture.Async
+import io.element.android.libraries.architecture.Presenter
+import io.element.android.libraries.matrix.api.verification.SessionVerificationService
+import javax.inject.Inject
+
+class VerifySelfSessionPresenter @Inject constructor(
+    private val sessionVerificationService: SessionVerificationService,
+) : Presenter<VerifySelfSessionState> {
+
+    init {
+        // Force reset, just in case the service was left in a broken state
+        sessionVerificationService.reset()
+    }
+
+    @Composable
+    override fun present(): VerifySelfSessionState {
+        val coroutineScope = rememberCoroutineScope()
+        val stateMachine = remember { VerifySelfSessionStateMachine(coroutineScope, sessionVerificationService) }
+
+        // Create the new view state from the StateMachine state
+        val stateMachineCurrentState by stateMachine.state.collectAsState()
+        val verificationState by remember {
+            derivedStateOf {
```

Comment:
`derivedStateOf` is used when the result of the computation changes less often than the input (here `stateMachineCurrentState`). is it the case here? 
Also maybe extract the mapping in a method

Reply:
I think that's the case here, we have several 'loading' states that happen very close in time and I think I saw way more recompositions than necessary.

I will extract the mapping function 👍 .